### ISSUE TITLE
typedef top level declaration

### ISSUE THREAD

**Issue by andrewrk** [andrewrk/zig#95]
``` zig
type foo = expr;
```

Usually you can simply use `const int = i32` for a "typedef". But in that case, the compiler sees those types as exactly the same thing.

Typedefs are for when you want the compiler to enforce that the types are distinct, and you have to do an explicit cast to convert the typedef'd thing to another type. Example:

``` zig
type inches = f32;
type meters = f32;
const x: inches = 13;
const y: meters = x; // error, type mismatch.
```

Another example is if you import a .h file with something like this:

``` c
struct Foo;
struct Bar;

void do_something_with_foo(struct Foo *foo);
void do_something_with_bar(struct Bar *bar);
```

If we import this as:

``` zig
const Foo = u8;
const Bar = u8;
fn do_something_with_foo(foo: &Foo);
fn do_something_with_bar(bar: &Bar);
```

It will work, but you can give the wrong type to a function and it will accept it, because really, both functions' first parameter has type `&u8`. However if we import it as:

``` zig
type Foo = u8;
type Bar = u8;
fn do_something_with_foo(foo: &Foo);
fn do_something_with_bar(bar: &Bar);
```

Now the compiler catches it if you accidentally supply the wrong type to the function.

Also, you can `export` the declaration to generate it in an .h file.

**Comment by PavelVozenilek** [comment]
Strong typedefs (I think this is the term) need operations (+ for integers, access by index for arrays, etc).  Will these operations get automatically cloned too?

In language Nim it is [very easy to create unique type](http://nim-lang.org/docs/manual.html#types-distinct-type):

    type
      Dollar = distinct int
      Euro = distinct int

Unfortunately no operation is defined by default on such new types and this makes them less than ideal. One has to type *a lot of code* (even assignement) to make them useful again.

**Comment by andrewrk** [comment]
I think yes the operations will get automatically cloned too. If Foo typedefs to a i32 then you can do i32 operations on it. The only thing you can't do is give a Foo when you were expecting another type, including i32.

**Comment by PavelVozenilek** [comment]
Probably "operations defined in the language" should be cloned automatically. E.g. arithmetic for integers, but not ```foo(int32)``` written by the user.

Nim uses special magic ```{.borrow.}``` to express that certain functionality of old (parent) type should be also cloned. IMO it is rather clumsy, lot of noise.

**Comment by andrewrk** [comment]
solved by #314
